Sources/SwiftProductGenerator.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  SwiftProductGenerator.swift
+//
+//
+//  Created by Øyvind Grimnes on 09/09/15.
+//
+//
+
+
+/**
+A cartesian product generator
+
+:discussion:    Generates the cartesian product of collections, or repetitions
+                of a collection. The rightmost element is advanced every
+                iteration which ensures that if the input is sorted, the
+                output will be sorted as well.
+*/
+public struct Product<T, C: Collection>: IteratorProtocol, Sequence
+    where C.Iterator.Element == T
+{
+
+    /// Private variable to manage where to get next element in each pool
+    private var indices : [C.Index]
+    /// Pool containing all the collections to be combined
+    private var pools   : [C]
+    /// Terminate the generation on completion
+    private var done    : Bool = false
+
+
+    /**
+    Initiates the generator with an array of collections to be combined.
+
+    :param:        collections    array containing the collections to be
+                                  combined
+    */
+    public init(_ collections: [C]) {
+        self.pools   = collections
+        self.indices = collections.map{ $0.startIndex }
+        self.done    = pools.reduce(true, { $0 && $1.count == 0 })
+    }
+
+    /**
+    Initiates the generator with a collection repeated 'count' times.
+
+    :param:        repeating      collection to be combined with itself
+    :param:        count          number of times to repeat the collection
+    */
+
+    public init(repeating collection: C, count: Int) {
+        precondition(count >= 0, "count must be >= 0")
+        self.init([C](repeating: collection, count: count))
+    }
+
+
+    /**
+    Generates the next element and returns it.
+    */
+    public mutating func next() -> [T]? {
+        if done {
+            return nil
+        }
+
+        let element = self.pools.enumerated().map {
+            $1[ self.indices[$0] ]
+        }
```

Comment:
I generally prefer having explicit `self` everywhere (the same way I prefer explicitly using `this` in c++) because I think it makes it clearer that I refer to a property and not a local variables. You had also some "unnecessary" `self` references in your original code, so I thought you'd be okay with.